CompilerSource/Proto2ES.cpp

```diff
@@ -0,0 +1,737 @@
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include "Proto2ES.h"
+
+#include <unordered_map>
+#include <vector>
+
+Sprite AddSprite(const buffers::resources::Sprite& spr);
+SubImage AddSubImage(const std::string fPath);
+Sound AddSound(const buffers::resources::Sound& snd);
+Background AddBackground(const buffers::resources::Background& bkg);
+Path AddPath(const buffers::resources::Path& pth);
+PathPoint AddPathPoint(const buffers::resources::Path::Point& pnt);
+Script AddScript(const buffers::resources::Script& scr);
+Shader AddShader(const buffers::resources::Shader& shr);
+Font AddFont(const buffers::resources::Font& fnt);
+Timeline AddTimeline(buffers::resources::Timeline* tml, buffers::Project* protobuf);
+GmObject AddObject(buffers::resources::Object* obj, buffers::Project* protobuf);
+Room AddRoom(const buffers::resources::Room& rmn, buffers::Project* protobuf);
+Instance AddInstance(const buffers::resources::Room::Instance& inst, buffers::Project* protobuf);
+Tile AddTile(const buffers::resources::Room::Tile& tile, buffers::Project* protobuf);
+View AddView(const buffers::resources::Room::View& view, buffers::Project* protobuf);
+BackgroundDef AddRoomBackground(const buffers::resources::Room::Background& bkg, buffers::Project* protobuf);
+
+template <class T>
+int Name2Id(const ::google::protobuf::RepeatedPtrField< T >& group, std::string name) {
+  for (auto& message : group) {
+    const google::protobuf::Descriptor *desc = message.GetDescriptor();
+    const google::protobuf::Reflection *refl = message.GetReflection();
+
+    const google::protobuf::FieldDescriptor *nameField = desc->FindFieldByName("name");
+    if (refl->GetString(message, nameField) == name) {
+      const google::protobuf::FieldDescriptor *idField = desc->FindFieldByName("id");
+      return refl->GetInt32(message, idField);
+    }
+  }
+  return -1;
+}
+
+inline std::string string_replace_all(std::string str, std::string substr, std::string nstr)
+{
+  size_t pos = 0;
+  while ((pos = str.find(substr, pos)) != std::string::npos)
+  {
+    str.replace(pos, substr.length(), nstr);
+    pos += nstr.length();
+  }
+  return str;
+}
+
+unsigned int BGR2RGBA(unsigned int x) {
+  return (((x & 0xFF0000) >> 16) | (x & 0xFF00) | ((x & 0xFF) << 16)) << 8 | 0xFF;
+}
+
+std::string Argument2Code(const buffers::resources::Argument& arg) {
+  using buffers::resources::ArgumentKind;
+  std::string val = arg.string();
+
+  switch (arg.kind()) {
+    case ArgumentKind::ARG_BOTH:
+      // treat as literal if starts with quote (")
+      if (val[0] == '"' || val[0] == '\'') return val;
+      // else fall through
+    case ArgumentKind::ARG_STRING:
+      return '\"' + string_replace_all(string_replace_all(val, "\\", "\\\\"), "\"", "\"+'\"'+\"") + '\"';
```

Comment:
I literally copied and pasted it from the compiler. I was afraid to include it probably because I didn't want other CompilerSource code bloating this layer. But I just realized this proto2es layer is compiled into CompilerSource anyway, so ok, I'll just include it.